File: tests/ccl_test_distances.c
#include "ccl.h"
#include "ctest.h"
#include <stdio.h>
#include <math.h>

// The tolerance in chi for all the
#define DISTANCES_TOLERANCE 1.0e-4

CTEST_DATA(distances) {
  double Omega_c;
  double Omega_b;
  double h;
  double A_s;
  double n_s;
  double N_nu_rel;
  double N_nu_mass;
  double mnu;
  double Omega_v[5];
  double Omega_k[5];
  double w_0[5];
  double w_a[5];
  
  double z[6];
  double chi[5][6];
  double dm[5][6];
};

// Read the fixed format file containing all the radial comoving
// distance benchmarks
static void read_chi_test_file(double z[6], double chi[5][6])
{
  //Distances are in Mpc/h
  FILE * f = fopen("./tests/benchmark/chi_model1-5.txt", "r");
  ASSERT_NOT_NULL(f);
  
  // Ignore header line
  char str[1024];
  fgets(str, 1024, f);
  
  // File is fixed format - five rows and six columns
  for (int i=0; i<6; i++) {
    int count = fscanf(f, "%le %le %le %le %le %le\n", &z[i],
               &chi[0][i], &chi[1][i], &chi[2][i], &chi[3][i], &chi[4][i]);
    // Check that all the stuff in the benchmark is there
    ASSERT_EQUAL(6, count);
  }
  fclose(f);
}

static void read_dm_test_file(double z[6], double dm[5][6])
{
  //Distances are in Mpc/h
  FILE * f = fopen("./tests/benchmark/dm_model1-5.txt", "r");
  ASSERT_NOT_NULL(f);

  // Ignore header line
  char str[1024];
  fgets(str, 1024, f);

  // File is fixed format - five rows and six columns
  for (int i=0; i<6; i++) {
    int count = fscanf(f, "%le %le %le %le %le %le\n", &z[i],
                       &dm[0][i], &dm[1][i], &dm[2][i], &dm[3][i], &dm[4][i]);
    // Check that all the stuff in the benchmark is there
    ASSERT_EQUAL(6, count);
  }
  fclose(f);
}

// Set up the cosmological parameters to be used in each of the
// models
CTEST_SETUP(distances) {
  // Values that are the same for all 5 models
  data->Omega_c = 0.25;
  data->Omega_b = 0.05;
  data->h = 0.7;
  data->A_s = 2.1e-9;
  data->n_s = 0.96;
  data->N_nu_rel=0;
  data->N_nu_mass=0;
  data->mnu=0;

  // Values that are different for the different models
  double Omega_v[5] = {  0.7,  0.7,  0.7,  0.65, 0.75 };
  double w_0[5]     = { -1.0, -0.9, -0.9, -0.9, -0.9  };
  double w_a[5]     = {  0.0,  0.0,  0.1,  0.1,  0.1  };

  // Fill in the values from these constant arrays.
  for (int i=0; i<5; i++) {
    data->Omega_v[i] = Omega_v[i];
    data->w_0[i]     = w_0[i];
    data->w_a[i]     = w_a[i];
    data->Omega_k[i] = 1.0 - data->Omega_c - data->Omega_b - data->Omega_v[i];
  }

  // The file of benchmark data.
  read_dm_test_file(data->z, data->dm);
  read_chi_test_file(data->z, data->chi);
}

static void compare_distances(int model, struct distances_data * data)
{
  int status=0;
  // Make the parameter set from the input data
  // Values of some parameters depend on the model index
  ccl_parameters params = ccl_parameters_create(data->Omega_c, data->Omega_b, data->Omega_k[model],
						data->N_nu_rel, data->N_nu_mass, data->mnu,
						data->w_0[model], data->w_a[model],
						data->h, data->A_s, data->n_s,-1,NULL,NULL, &status);
  
  params.Omega_g=0;
  
  // Make a cosmology object from the parameters with the default configuration
  ccl_cosmology * cosmo = ccl_cosmology_create(params, default_config);
  ASSERT_NOT_NULL(cosmo);
  
  // Compare to benchmark data
  for (int j=0; j<6; j++) {
    double a = 1/(1.+data->z[j]);
    double chi_ij=ccl_comoving_radial_distance(cosmo,a, &status)*data->h;
    if (status) printf("%s\n",cosmo->status_message);
    double absolute_tolerance = DISTANCES_TOLERANCE*data->chi[model][j];
    if (fabs(absolute_tolerance)<1e-12) absolute_tolerance = 1e-12;
    ASSERT_DBL_NEAR_TOL(data->chi[model][j], chi_ij, absolute_tolerance);

    if(a!=1) {  //skip this test for a=1 since it will raise an error
        double dm_ij=ccl_distance_modulus(cosmo,a, &status);
        if (status) printf("%s\n",cosmo->status_message);
        //NOTE tolerances are different!
        absolute_tolerance = 10*DISTANCES_TOLERANCE*data->dm[model][j];
        if (fabs(absolute_tolerance)<1e-4) absolute_tolerance = 1e-4;
        ASSERT_DBL_NEAR_TOL(data->dm[model][j], dm_ij, absolute_tolerance);
    }
  }
  
  ccl_cosmology_free(cosmo);
}

CTEST2(distances, model_1) {
  int model = 0;
  compare_distances(model, data);
}

CTEST2(distances, model_2) {
  int model = 1;
  compare_distances(model, data);
}

CTEST2(distances, model_3) {
  int model = 2;
  compare_distances(model, data);
}

CTEST2(distances, model_4) {
  int model = 3;
  compare_distances(model, data);
}

CTEST2(distances, model_5) {
  int model = 4;
  compare_distances(model, data);
}

File: tests/ccl_test_distances.py
import numpy as np
from numpy.testing import assert_allclose, run_module_suite
import pyccl as ccl
from os.path import dirname,join,abspath
# Set tolerances
DISTANCES_TOLERANCE = 1e-4

# Set up the cosmological parameters to be used in each of the models
# Values that are the same for all 5 models
Omega_c = 0.25
Omega_b = 0.05
N_nu_rel = 0.
h = 0.7
A_s = 2.1e-9
n_s = 0.96

# Values that are different for the different models
Omega_v_vals = np.array([0.7, 0.7, 0.7, 0.65, 0.75])
w0_vals = np.array([-1.0, -0.9, -0.9, -0.9, -0.9])
wa_vals = np.array([0.0, 0.0, 0.1, 0.1, 0.1])
path = dirname(abspath(__file__))
def read_chi_test_file():
    """
    Read the file containing all the radial comoving distance benchmarks 
    (distances are in Mpc/h)
    """
    # Load data from file
    dat = np.genfromtxt(join(path,"benchmark/chi_model1-5.txt")).T
    assert(dat.shape == (6,6))
    
    # Split into redshift column and chi(z) columns
    z = dat[0]
    chi = dat[1:]
    return z, chi

def read_chi_hiz_test_file():
    """
    Read the file containing all the radial comoving distance benchmarks 
    (distances are in Mpc/h)
    """
    # Load data from file
    dat = np.genfromtxt(join(path,"benchmark/chi_hiz_model1-3.txt")).T
    assert(dat.shape == (4,7))
    
    # Split into redshift column and chi(z) columns
    z = dat[0]
    chi = dat[1:]
    return z, chi

def read_dm_test_file():
    """
    Read the file containing all the distance modulus benchmarks 
    """
    # Load data from file
    dat = np.genfromtxt(join(path,"benchmark/dm_model1-5.txt")).T
    assert(dat.shape == (6,6))

    # Split into redshift column and chi(z) columns
    z = dat[0]
    dm = dat[1:]
    return z, dm

# Set-up test data
z, chi = read_chi_test_file()
zhi, chi_hiz = read_chi_hiz_test_file()
_, dm = read_dm_test_file()

def compare_distances(z, chi_bench,dm_bench, Omega_v, w0, wa):
    """
    Compare distances calculated by pyccl with the distances in the benchmark 
    file.
    This test is only valid when radiation is explicitly set to 0.
    """
    # Set Omega_K in a consistent way
    Omega_k = 1.0 - Omega_c - Omega_b - Omega_v    
    
    # Create new Parameters and Cosmology objects
    p = ccl.Parameters(Omega_c=Omega_c, Omega_b=Omega_b, N_nu_rel=N_nu_rel, 
                       h=h, A_s=A_s, n_s=n_s, Omega_k=Omega_k,
                       w0=w0, wa=wa)
    p.parameters.Omega_g = 0. # Hack to set to same value used for benchmarks
    cosmo = ccl.Cosmology(p)
    
    # Calculate distance using pyccl
    a = 1. / (1. + z)
    chi = ccl.comoving_radial_distance(cosmo, a) * h
    # Compare to benchmark data
    assert_allclose(chi, chi_bench, atol=1e-12, rtol=DISTANCES_TOLERANCE)

    #compare distance moudli where a!=1
    a_not_one = (a!=1).nonzero()
    dm = ccl.distance_modulus(cosmo,a[a_not_one])

    assert_allclose(dm, dm_bench[a_not_one], atol=1e-3, rtol = DISTANCES_TOLERANCE*10)

def compare_distances_hiz(z, chi_bench, Omega_v, w0, wa):
    """
    Compare distances calculated by pyccl with the distances in the benchmark 
    file.
    This test is only valid when radiation is explicitly set to 0.
    """
    # Set Omega_K in a consistent way
    Omega_k = 1.0 - Omega_c - Omega_b - Omega_v    
    
    # Create new Parameters and Cosmology objects
    p = ccl.Parameters(Omega_c=Omega_c, Omega_b=Omega_b, N_nu_rel=N_nu_rel, 
                       h=h, A_s=A_s, n_s=n_s, Omega_k=Omega_k,
                       w0=w0, wa=wa)
    p.parameters.Omega_g = 0. # Hack to set to same value used for benchmarks
    cosmo = ccl.Cosmology(p)
    
    # Calculate distance using pyccl
    a = 1. / (1. + z)
    chi = ccl.comoving_radial_distance(cosmo, a) * h
    # Compare to benchmark data
    assert_allclose(chi, chi_bench, atol=1e-12, rtol=DISTANCES_TOLERANCE)
    
def test_distance_model_0():
    i = 0
    compare_distances(z, chi[i],dm[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_distance_model_1():
    i = 1
    compare_distances(z, chi[i],dm[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_distance_model_2():
    i = 2
    compare_distances(z, chi[i],dm[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_distance_model_3():
    i = 3
    compare_distances(z, chi[i],dm[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_distance_model_4():
    i = 4
    compare_distances(z, chi[i],dm[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_distance_hiz_model_0():
    i = 0
    compare_distances_hiz(zhi, chi_hiz[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_distance_hiz_model_1():
    i = 1
    compare_distances_hiz(zhi, chi_hiz[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])

def test_distance_hiz_model_2():
    i = 2
    compare_distances_hiz(zhi, chi_hiz[i], Omega_v_vals[i], w0_vals[i], wa_vals[i])
    

if __name__ == "__main__":
    run_module_suite()
